### PR TITLE
Support sqlite3 gem

### PR DESCRIPTION
**Pull request by nirvdrum**

### CODE REVIEW

File: lib/cext/preprocess.rb [hunk 1]
```diff
@@ -783,6 +783,103 @@ def tuple_new_patch(ctx, slf)
...
+        replacement: 'VALUE self = rb_tr_managed_from_handle(ctx);'
+      },
+      { # busy_handler
+        match: 'ctx->db, NIL_P(block) ? NULL : rb_sqlite3_busy_handler, (void *)self);',
+        replacement: 'ctx->db, NIL_P(block) ? NULL : rb_sqlite3_busy_handler, rb_tr_handle_for_managed_leaking(self));'
+      },
+      { # rb_sqlite3_func
+        match: 'VALUE callable = (VALUE)sqlite3_user_data(ctx);',
+        replacement: 'VALUE callable = rb_tr_managed_from_handle(sqlite3_user_data(ctx));'
+      },
```
Comment (bjfish) [inline lib/cext/preprocess.rb]: Is this one repeated 2 more times below?

File: lib/cext/preprocess.rb [hunk 2]
```diff
@@ -783,6 +783,103 @@ def tuple_new_patch(ctx, slf)
...
+        match: 'comparator = (VALUE)ctx;',
+        replacement: 'comparator = rb_tr_managed_from_handle(ctx);'
+      },
+      { # collation
+        match: '(void *)comparator,',
+        replacement: 'rb_tr_handle_for_managed_leaking(comparator),'
+      },
+      { # enc_cb
+        match: 'rb_iv_set(self, "@encoding", rb_enc_from_encoding(e));',
+        replacement: 'rb_iv_set(rb_tr_managed_from_handle(_self), "@encoding", rb_enc_from_encoding(e));'
```
Comment (bjfish) [inline lib/cext/preprocess.rb]: I there supposed to be underscore here? `_self`

File: lib/cext/preprocess.rb [hunk 3]
```diff
@@ -783,6 +783,103 @@ def tuple_new_patch(ctx, slf)
...
+  },
+  'database.c' => {
+    gem: 'sqlite3',
+    patches: [
+      { # tracefunc
+        match: 'VALUE self = (VALUE)data;',
+        replacement: 'VALUE self = rb_tr_managed_from_handle(data);'
+      },
+      { # trace
+        match: 'sqlite3_trace(ctx->db, NIL_P(block) ? NULL : tracefunc, (void *)self);',
```
Comment (bjfish) [inline lib/cext/preprocess.rb]: Consider a regex replacement for `(void *)self`, this one appears to repeat a few times.

File: src/main/java/org/truffleruby/core/range/RangeNodes.java [hunk 4]
```diff
@@ -76,6 +77,13 @@ public DynamicObject map(DynamicObject range, DynamicObject block,
             return createArray(arrayBuilder.finish(store, length), length);
         }
 
+        @Specialization(guards = "isObjectRange(range)")
+        public Object mapObject(VirtualFrame frame, DynamicObject range, DynamicObject block,
+                                @Cached("new()") SnippetNode snippetNode) {
+            return snippetNode.execute(frame, "ret = []; range.each { |x| ret << block.call(x) }; ret",
+                    "range", range, "block", block);
```
Comment (bjfish) [inline src/main/java/org/truffleruby/core/range/RangeNodes.java]: Is this the same as `#map`?

File: src/main/java/org/truffleruby/core/range/RangeNodes.java [hunk 5]
```diff
@@ -76,6 +77,13 @@ public DynamicObject map(DynamicObject range, DynamicObject block,
...
+        }
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/range/RangeNodes.java]: Could you make this a primitive and have a fallback specialization so this code is not hidden in a Snippet but lives in core/range.rb? This should also handle `long` ranges trivially.

File: lib/cext/preprocess.rb [hunk 6]
```diff
@@ -834,7 +834,7 @@ def patch(file, contents, directory)
                 directory.end_with?(File.join(patched_file[:gem], 'ext', patched_file[:ext_dir]))
               else
                 regexp = /^#{Regexp.escape(patched_file[:gem])}\b/
-                directory.split('/').last(2).any? { |part| part =~ regexp }
+                directory.split('/').last(2).any? { |part| part =~ regexp } || file.split('/').last(2).any? { |part| part =~ regexp }
```
Comment (eregon) [inline lib/cext/preprocess.rb]: Maybe we should just match with `File.expand_path(file).end_with?("/#{key}")` but that would require iterating PATCHED_FILES's keys (likely OK here).
We could also consistently use a fixed number of path components like 2, then we can do a simple lookup.

File: lib/cext/preprocess.rb [hunk 7]
```diff
@@ -783,6 +783,103 @@ def tuple_new_patch(ctx, slf)
...
+        match: 'ctx->db, NIL_P(block) ? NULL : rb_sqlite3_busy_handler, (void *)self);',
+        replacement: 'ctx->db, NIL_P(block) ? NULL : rb_sqlite3_busy_handler, rb_tr_handle_for_managed_leaking(self));'
+      },
+      { # rb_sqlite3_func
+        match: 'VALUE callable = (VALUE)sqlite3_user_data(ctx);',
+        replacement: 'VALUE callable = rb_tr_managed_from_handle(sqlite3_user_data(ctx));'
+      },
+      { # define_function_with_flags
+        match: 'NUM2INT(flags),',
+        replacement: 'flags,'
```
Comment (eregon) [inline lib/cext/preprocess.rb]: Why is this needed?

File: lib/cext/preprocess.rb (hunk 3 above)
Comment (nirvdrum) [reply lib/cext/preprocess.rb]: I really prefer matching the exact string. I think the regexp matches we do are major headache waiting to happen.

File: lib/cext/preprocess.rb (hunk 2 above)
Comment (nirvdrum) [reply lib/cext/preprocess.rb]: Yes. That's the parameter name to the `enc_cb` function. I could move the managed handle creation elsewhere though and make the name `self`

File: lib/cext/preprocess.rb (hunk 7 above)
Comment (nirvdrum) [reply lib/cext/preprocess.rb]: Evidently it's not. In a fit of debugging I thought it helped make progress, but I just removed it and everything seems to work.

File: lib/cext/preprocess.rb (hunk 2 above)
Comment (nirvdrum) [reply lib/cext/preprocess.rb]: I've updated the patch to move the call location and better match the other patches here.

File: src/main/java/org/truffleruby/core/range/RangeNodes.java (hunk 5 above)
Comment (nirvdrum) [reply src/main/java/org/truffleruby/core/range/RangeNodes.java]: Check 5232f8afd56b3e24fec71b20342b7024e6732811 and see if that's what you had in mind, please.

File: lib/cext/preprocess.rb (hunk 6 above)
Comment (nirvdrum) [reply lib/cext/preprocess.rb]: I agree this needs to be cleaned up. The basic assumption made when this pattern matching system was set hasn't held. The naming conventions used are much more varied. But, let's handle that in a separate PR if you don't mind.

File: lib/cext/preprocess.rb (hunk 3 above)
Comment (pitr-ch) [reply lib/cext/preprocess.rb]: yeah we should think about replacing the regex with diff. It would be more exact and it also little bit more understandable in the source code what is being changed.

Comment (eregon) [reply lib/cext/preprocess.rb]: A diff would fail very quickly to apply with any change around though.
I'd rather focus on not needing these patches.

File: src/main/java/org/truffleruby/core/range/RangeNodes.java (hunk 5 above)
Comment (eregon) [reply src/main/java/org/truffleruby/core/range/RangeNodes.java]: Yes it is, thanks.

File: lib/cext/preprocess.rb (hunk 6 above)
Comment (eregon) [reply lib/cext/preprocess.rb]: Sure.

File: src/main/java/org/truffleruby/core/range/RangeNodes.java [hunk 8]
```diff
@@ -76,6 +79,19 @@ public DynamicObject map(DynamicObject range, DynamicObject block,
...
+
+        private Object dispatch(DynamicObject block, Object... argumentsObjects) {
+            if (yieldNode == null) {
+                CompilerDirectives.transferToInterpreterAndInvalidate();
+                yieldNode = insert(new YieldNode());
+            }
+
+            return yieldNode.dispatch(block, argumentsObjects);
+        }
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/range/RangeNodes.java]: You could use `@Cached("new()")` here to make it much shorter.

File: src/main/ruby/core/range.rb [hunk 9]
```diff
@@ -326,6 +326,14 @@ def size
     end
   end
 
+  def collect
+    Truffle.primitive :range_integer_map
+
+    super
+  end
+
+  alias_method :map, :collect
```
Comment (eregon) [inline src/main/ruby/core/range.rb]: I like `alias_method` without vertical space so it's clear it forms a group with the `def` above.

File: test/truffle/cexts/sqlite3/test.rb [hunk 10]
```diff
@@ -0,0 +1,9 @@
+require 'sqlite3'
+
+db = SQLite3::Database.new(':memory:')
+
+db.execute( 'CREATE TABLE test(id INTEGER, name VARCHAR(20) NOT NULL)')
```
Comment (eregon) [inline test/truffle/cexts/sqlite3/test.rb]: Extra space here.

File: src/main/ruby/core/range.rb (hunk 9 above)
Comment (nirvdrum) [reply src/main/ruby/core/range.rb]: I just copied the style used in enumerable.rb, but easy enough to change.

File: src/main/java/org/truffleruby/core/range/RangeNodes.java (hunk 8 above)
Comment (nirvdrum) [reply src/main/java/org/truffleruby/core/range/RangeNodes.java]: Thanks. I didn't realize that would work. And all other usages of `YieldNode` make it a field.
